packages/@lwc/engine/src/framework/hooks.ts

```diff
@@ -163,13 +163,14 @@ export function allocateChildrenHook(vnode: VCustomElement) {
     const elm = vnode.elm as HTMLElement;
     const vm = getCustomElementVM(elm);
     const { children } = vnode;
-    vm.aChildren = children;
     if (isTrue(useSyntheticShadow)) {
         // slow path
         allocateInSlot(vm, children);
         // every child vnode is now allocated, and the host should receive none directly, it receives them via the shadow!
         vnode.children = EmptyArray;
     }
+    // storing the new children collection for disconnection purposes
+    vm.aChildren = children;
```

Comment:
this can't be done before `allocateInSlot` because inside that routine we might need to access the old children which is accessible from `aChildren`.